_posts/2022-08-04-calldata-tuple-reencoding-head-overflow-bug.md

```diff
@@ -0,0 +1,219 @@
+    E.g. `bytes` or a struct containing a dynamically-sized array.
+
+Note that structs are represented as tuples in the ABI.
+Parameter and return value lists of external functions, events and errors are also implicitly treated as tuples.
+
+The bug can be triggered by explicit use of the `abi.encode...` family of functions as well as the ABI-encoding performed
+implicitly for external functions, events and errors by the compiler.
+
+The bug is independent of the use of the optimizer and affects both the legacy and the IR-based compilation pipelines.
+
+## Technical Details
+
+The bug is the result of overly eager cleanup performed by the compiler when copying calldata arrays to memory during ABI-encoding.
+Arrays in memory always occupy a multiple of 32 bytes and when the base type does not fill the whole word,
+the unused space is guaranteed to be zeroed and remain clean after all high-level Solidity operations.
+In the affected cases the compiler would emit code that incorrectly cleaned the end of an array stored in the last component of a tuple,
+zeroing 32 bytes belonging to the first dynamically-sized component of the tuple.
+
+### Aggressive Array Cleanup
+
+When the input for the ABI encoder comes from calldata, large parts of it can simply be copied to the output.
+The encoder, however, must ensure that the unused spaces between elements are actually
+[cleaned](https://docs.soliditylang.org/en/latest/internals/variable_cleanup.html).
+
+Depending on the types, more or less work may be needed and the encoder has several code paths for handling them.
+One of them was designed for (possibly multi-dimensional) arrays of simple value types that fill the whole word,
+i.e. `uint` and `bytes32`, as well as byte arrays, i.e. `bytes` and `string`.
+In this case, due to tight packing of array elements, only the area past the array may need cleanup and that
+area is never larger than 32 bytes.
+
+The cleanup was being performed using a single `mstore()` instruction, which always writes exactly 32 bytes.
+The cleanup was only ever required for byte arrays but was not considered harmful for `uint` and `bytes32`,
+because of the order in which the encoder writes tuple components.
+The space after any given tuple component was assumed to be safe to write because it would eventually be
+overwritten with the data belonging to the next component.
+
+In order to explain why the assumption turned out to be incorrect, let us take a closer look at
+the layout of an ABI-encoded tuple.
+
+### Layout and Encoding Order of Statically- and Dynamically-sized Tuples
+
+[The ABI encoding of a tuple](https://docs.soliditylang.org/en/latest/abi-spec.html#formal-specification-of-the-encoding)
+consists of two areas - the statically-sized *head* and the dynamically-sized *tail*.
+Every statically-sized component is placed directly in the head area while for dynamically-sized ones, the head contains only the offset
```

Comment:
I wonder how nitpicky we should be here... is ``bytes[1]`` "statically-sized" :-)? Because things like that are "dynamically encoded" and end up in the tail as well... but not sure if the distinction is helpful or only confusing...

Reply:
Do we have a precise definition of "dynamically-sized" somewhere? I've always been thinking of things like `bytes[1]` as dynamically sized. I.e. the data as a whole does not have a fixed size even if it some parts of it some static or it's a part of an array with a fixed number of elements.

Basically, the dynamically-sized stuff is contagious and pollutes everything it touches :)

Reply:
Yeah, I was mainly thinking in terms of ``Type::isDynamicallySized`` and ``Type::isDynamicallyEncoded`` in our code-base :-). (the former being false for ``bytes[1]``, the latter being true)
But we don't seem to use those terms in the docs actually...
https://docs.soliditylang.org/en/v0.8.15/abi-spec.html talks about "fixed-size" or "fixed-length" arrays and "non-fixed-size types" and of "dynamic" types (``bytes[1]`` would be a dynamic fixed-size array according to that...)...

Maybe it's just fine to keep it at the current phrasing for the blog post in any case, just wanted to point out that there is some potential for ambiguity here.

Reply:
Hmm... I always had the impression that @chriseth was careful to always refer to them as "statically-sized" vs "dynamically-sized" over any other terms, including "dynamic" and "static". Maybe one day we should add a glossary or something to the docs to make sure we're all using the same terms :)

Anyway in the post I was paying attention to only use "statically-sized" and "dynamically-sized" and to use them consistently according so I'll leave that as is now.

Reply:
Well, for arrays the terms are ``statically-sized`` and ``dynamically-sized`` - but ``bytes[1]`` counts as statically-sized and the usual phrase for it still being encode via tail pointers is ``dynamically encoded`` (also occurs in the bug list on a few occasions like that)